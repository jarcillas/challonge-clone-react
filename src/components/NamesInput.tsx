import { useState, type ChangeEvent } from 'react';

import { generateInitialRounds } from '../helpers/generateInitialRounds';

type NameInputProps = {
  setRoundsData: Function;
  showInput: boolean;
  setShowNamesInput: Function;
};

const NamesInput = ({
  setRoundsData,
  showInput,
  setShowNamesInput,
}: NameInputProps) => {
  const defaultPlayers = [...Array(14).keys()].map((n) => `Player ${n + 1}`);

  const [namesInput, setNamesInput] = useState(defaultPlayers.join('\n'));

  const onNamesInputChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    setNamesInput(e.target.value);
  };

  const onCreateBracketClick = (e: React.MouseEvent<HTMLInputElement>) => {
    e.preventDefault();
    const playerList = [...new Set(namesInput.split('\n'))]; // generate a list of players from the names input

    // Don't generate a bracket if there are less than 2 players
    if (playerList.length < 2) {
      return;
    }
    setRoundsData(generateInitialRounds(playerList));
  };

  const onToggleButtonClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setShowNamesInput(!showInput);
  };

  return (
    <>
      <form id="input-container">
        <h4 hidden={!showInput} id="input-heading">
          Please enter the names of the players/teams ordered by their seed:
        </h4>
        <textarea
          onChange={onNamesInputChange}
          hidden={!showInput}
          id="name-input"
          cols={30}
          rows={20}
          value={namesInput}
          spellCheck={false}
        ></textarea>
        <input
          hidden={!showInput}
          type="submit"
          value="Create bracket"
          id="create-bracket-button"
          onClick={onCreateBracketClick}
        />
        <button onClick={onToggleButtonClick}>
          {showInput ? 'Hide' : 'Show'}
        </button>
      </form>
    </>
  );
};

export { NamesInput };
