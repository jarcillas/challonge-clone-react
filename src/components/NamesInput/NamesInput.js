import { useState } from 'react';

const NamesInput = ({ setRoundsData }) => {
  const defaultPlayers = [...Array(20).keys()].map((n) => `Player ${n + 1}`);

  const [namesInput, setNamesInput] = useState(defaultPlayers.join('\n'));

  const onNamesInputChange = (e) => {
    setNamesInput(e.target.value);
  };

  const onCreateBracketClick = (e) => {
    e.preventDefault();
    const playerList = [...new Set(namesInput.split('\n'))]; // generate a list of players from the names input
    const playerCount = playerList.length;
    if (playerCount < 2) {
      return;
    }
    const roundCount = Math.ceil(Math.log2(playerCount)); // calculate total number of rounds needed
    let initialRoundsData = []; // initialize initialRoundsData
    const byeCount = 2 ** roundCount - playerCount; // calculate total number of byes in the 1st round
    const playerByeList = [...playerList, ...Array(byeCount).fill('Bye')]; // create a player list (ordered by seeding) for the 1st round including all the byes

    initialRoundsData.push(
      generateFirstRound(roundCount).map((num) => playerByeList[num - 1])
    );

    if (roundCount > 1) {
      initialRoundsData.push(
        generateSecondRound(initialRoundsData[0], roundCount)
      );
    }

    if (roundCount > 2) {
      initialRoundsData = [
        ...initialRoundsData,
        ...generateOtherRounds(roundCount),
      ];
    }

    setRoundsData(initialRoundsData);
  };

  // helper function to generate the 1st round seeding matchups recursively. Seeds start from 1.
  const generateFirstRound = (roundCount) => {
    if (roundCount === 1) {
      return [1, 2];
    } else {
      let tempArray = [];
      for (let num of generateFirstRound(roundCount - 1)) {
        tempArray.push(num);
        tempArray.push(2 ** roundCount - num + 1);
      }
      return tempArray;
    }
  };

  // helper function to generate the 2nd round that advances players facing byes in 1st round
  const generateSecondRound = (round1Data, roundCount) => {
    let round2 = [];
    for (let k = 0; k < 2 ** (roundCount - 1); k++) {
      if (round1Data[2 * k + 1] === 'Bye') {
        round2.push(round1Data[2 * k]);
      } else {
        round2.push(null);
      }
    }
    return round2;
  };

  // helper function to generate the rest of the rounds if roundCount > 2
  const generateOtherRounds = (roundCount) => {
    let otherRoundsData = [];
    for (let j = 2; j < roundCount; j++) {
      let roundArray = Array(2 ** (roundCount - j)).fill(null);
      otherRoundsData.push(roundArray);
    }

    return otherRoundsData;
  };

  return (
    <>
      <form id="input-container">
        <h4 id="input-heading">
          Please enter the names of the players/teams ordered by their seed:
        </h4>
        <textarea
          onChange={onNamesInputChange}
          id="name-input"
          cols="30"
          rows="20"
          value={namesInput}
        ></textarea>
        <input
          type="submit"
          value="Create bracket"
          id="create-bracket-button"
          onClick={onCreateBracketClick}
        />
      </form>
    </>
  );
};

export default NamesInput;
