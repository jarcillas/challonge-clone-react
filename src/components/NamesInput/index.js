import NamesInput from './NamesInput';

export default NamesInput;
