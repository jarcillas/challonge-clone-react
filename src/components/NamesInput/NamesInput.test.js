import renderer from 'react-test-renderer';

import NamesInput from './Namesinput';

it('renders correctly', () => {
  const component = renderer.create(<NamesInput />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
