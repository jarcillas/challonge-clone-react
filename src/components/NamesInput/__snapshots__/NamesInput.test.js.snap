// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  id="input-container"
>
  <h4
    hidden={true}
    id="input-heading"
  >
    Please enter the names of the players/teams ordered by their seed:
  </h4>
  <textarea
    cols="30"
    hidden={true}
    id="name-input"
    onChange={[Function]}
    rows="20"
    spellCheck={false}
    value="Player 1
Player 2
Player 3
Player 4
Player 5
Player 6
Player 7
Player 8
Player 9
Player 10
Player 11
Player 12
Player 13
Player 14"
  />
  <input
    hidden={true}
    id="create-bracket-button"
    onClick={[Function]}
    type="submit"
    value="Create bracket"
  />
  <button
    onClick={[Function]}
  >
    Show
  </button>
</form>
`;
